flink-staging/flink-table/src/main/scala/org/apache/flink/api/java/table/JavaBatchTranslator.scala

```diff
@@ -54,6 +56,26 @@ class JavaBatchTranslator extends PlanTranslator {
     Table(Root(rowDataSet, resultFields))
   }
 
+  override def createTable(tableSource: TableSource): Table = {
+    // a TableSource requires an ExecutionEnvironment
+    if (env == null) {
+      throw new InvalidProgramException("This operation requires an ExecutionEnvironment.")
+    }
+    if (tableSource.isInstanceOf[AdaptiveTableSource]) {
+      val adaptiveTs = tableSource.asInstanceOf[AdaptiveTableSource]
+      if (adaptiveTs.supportsResolvedFieldPushdown || adaptiveTs.supportsPredicatePushdown) {
+        Table(Root(adaptiveTs, adaptiveTs.getOutputFields()))
+      }
+      else {
+        Table(Root(adaptiveTs.createAdaptiveDataSet(env), adaptiveTs.getOutputFields()))
+      }
+    }
+    else {
+      val staticTs = tableSource.asInstanceOf[StaticTableSource]
+      createTable(staticTs.createStaticDataSet(env), staticTs.getOutputFieldNames().mkString(","))
+    }
```

Comment:
You can replace the if/else by:

```scala
tableSource match {
  case adaptive: AdaptiveTableSource if adaptive.supportsResolvedFieldPushdown && ... => ...

  case adaptive: AdaptiveTableSource => ...
  
  case static: StaticTableSource => ...

  case _ => throw new Exception("something unexpected")
}
```

if I'm not mistaken, this is just of the top of my head.